File: src/main/java/ru/arc/sync/base/SyncMessager.java
package ru.arc.sync.base;

import lombok.RequiredArgsConstructor;
import ru.arc.ARC;
import ru.arc.network.ChannelListener;
import ru.arc.network.RedisManager;

import static ru.arc.util.Logging.debug;


@RequiredArgsConstructor
public class SyncMessager<T extends SyncData> implements ChannelListener {

    final RedisManager redisManager;
    final String channel;
    final SyncRepo<T> repo;

    @Override
    public void consume(String channel, String message, String originServer) {
        if (originServer.equals(ARC.serverName)) return;
        T data = repo.gson.fromJson(message, repo.clazz);
        debug("Received data from channel: {} with data: {}", channel, data);
        repo.dataApplier.accept(data);
    }

    public void send(T data) {
        debug("Sending data to channel: {}", channel);
        redisManager.publish(channel, repo.gson.toJson(data));
    }
}

File: src/main/kotlin/ru/arc/sync/base/SyncMessager.kt
package ru.arc.sync.base

import ru.arc.ARC
import ru.arc.network.ChannelListener
import ru.arc.network.RedisManager
import ru.arc.util.Logging.debug

class SyncMessager<T : SyncData>(
    val redisManager: RedisManager,
    val channel: String,
    val repo: SyncRepo<T>,
) : ChannelListener {

    override fun consume(channel: String, message: String, originServer: String) {
        if (originServer == ARC.serverName) return
        val data: T = repo.gson.fromJson(message, repo.clazz)
        debug("Received data from channel: {} with data: {}", channel, data)
        repo.dataApplier(data)
    }

    fun send(data: T) {
        debug("Sending data to channel: {}", channel)
        redisManager.publish(channel, repo.gson.toJson(data))
    }
}
